### PR TITLE
feat: overload document service middlewares

### PR DESCRIPTION
**Pull request by Marc-Roig**
### What does it do?

Improvement on document service middleware. Now you can: 

```
// Middleware on any uid and any action
strapi.documents.use(async (ctx, next) => ...)
```

```
// Middleware on any action
strapi.documents('api::article.article').use(async (ctx, next) => ...)
```

These two use cases where not possible before.

I know @alexandrebodin  was not fully convinced about this approach, so let me know if we can provide other api alternatives :)!

### CODE REVIEW

File: packages/core/types/src/modules/documents/index.ts [hunk 1]
```diff
@@ -107,20 +112,32 @@ export type Service = {
...
+        action: TAction,
+        cb:
+          | Middleware.Middleware<Common.UID.ContentType, TAction>
+          | Middleware.Middleware<Common.UID.ContentType, TAction>[],
+        opts?: Middleware.Options
+      ) => Service)
+    | ((
+        cb: Middleware.Middleware<Common.UID.ContentType, keyof DocumentEngine>,
+        opts?: Middleware.Options
+      ) => Service);
```
Comment (Marc-Roig) [inline packages/core/types/src/modules/documents/index.ts:140]: @Convly I can not find a way to overload this properly, typings do not seem to get inferred correctly. Could we take a look at this tomorrow?  🙏

**Comment by derrickmehaffy** [comment]
These are basically document service lifecycles right?

**Comment by derrickmehaffy** [comment]
@Boegie19 another good one you've been asking about! :)

**Comment by Marc-Roig** [comment]
we are going to refactor the api a bit, closing this for now, expect news soon :)

**Pull request closed without merging**